### DISCUSSION TITLE
Open a desktop app from web browser?

### DISCUSSION THREAD

**Discussion by XiaoGen-Taiwan** [airsdk/Adobe-Runtime-Support#2119] [General]
Does anyone know how to open a desktop app (captive runtime bundle) from a web browser (as e.g. Discord).
I think before it was possible by enabling the `<allowBrowserInvocation/>` within the app descriptor file
and importing the AIRAliases.js within the webpage and call something like `launchApplication`.

**Comment by ajwfrost** [comment]
I think the invocation was always previously dealt with via the Flash Player .. and that's no longer possible.

So the only option now is to create a custom 'schema' and launch via that, which is how e.g. Zoom and the like would get their desktop applications opening from a URL, something like "zoom://abc.def" 

thanks

**Reply by XiaoGen-Taiwan** [reply]
Thanks a lot for your reply.
Good hint ... i will check more on registering custom url schemes ... maybe it can be done with an Installer app.

**Comment by itlancer** [comment]
@XiaoGen-Taiwan 
You need to register your custom URI scheme for that.

For **Windows**:
You need to add some registry values to end device. Better to make it in application installer.
You should create structure in registry something like below. Where "sample" in `HKEY_CLASSES_ROOT` is your custom URI scheme. After that you can open your application from browser clicking by link with `<a href="sample:...">`. Notice that some browsers should be restarted to start to use it.
```
HKEY_CLASSES_ROOT
   sample
      (Default) = "URL:Sample Protocol"
      URL Protocol = ""
      DefaultIcon
         (Default) = "sample.exe,1"
      shell
         open
            command
               (Default) = "C:\Program Files\Sample\sample.exe" "%1"
```
			   
More about it you can find for example here:
https://www.meziantou.net/registering-an-application-to-a-uri-scheme-using-net.htm

For **macOS**:
You should add to AIR application manifest:
```
<macOS>
	<InfoAdditions>
		<![CDATA[
			<key>CFBundleURLTypes</key>
			<array>
				<dict>
					<key>CFBundleTypeRole</key>
					<string>Shell</string>
					<key>CFBundleURLName</key>
					<string>your.app.URISchemeName</string>
					<key>CFBundleURLSchemes</key>
					<array>
						<string>sample</string>
					</array>
				</dict>
			</array>
		]]>
	</InfoAdditions>
</macOS>
```
where change `your.app.URISchemeName` and `sample` values to your own.
https://developer.apple.com/documentation/bundleresources/information_property_list/cfbundleurltypes/cfbundleurlschemes?language=objc

For **Android**:
You should add to AIR application manifest:
```
<android>
	<manifestAdditions>
		<![CDATA[
			<manifest xmlns:tools="http://schemas.android.com/tools">
				<application>
					<activity>
						<intent-filter>
							<action android:name="android.intent.action.VIEW" />
							<category android:name="android.intent.category.DEFAULT" />
							<category android:name="android.intent.category.BROWSABLE" />
							<data android:scheme="sample"/>
						</intent-filter>
					</activity>
				</application>
			</manifest>
		]]>
	</manifestAdditions>
</android>
```
where change `sample` value to your own.
https://developer.android.com/training/app-links/deep-linking

For **iOS**:
You should add to AIR application manifest:
```
<iPhone>
	<InfoAdditions>
		<![CDATA[
			<key>CFBundleURLTypes</key>
			<array>
				<dict>
					<key>CFBundleURLName</key>
					<string>your.app.URISchemeName</string>
					<key>CFBundleTypeRole</key>
					<string>Shell</string>
					<key>CFBundleURLSchemes</key>
					<array>
						<string>sample</string>
					</array>
				</dict>
			</array>
		]]>
	</InfoAdditions>
</iPhone>
```
where change `your.app.URISchemeName` and `sample` values to your own.
https://developer.apple.com/documentation/bundleresources/information_property_list/cfbundleurltypes/cfbundleurlschemes?language=objc

In all these cases when your application will be invoked such way you can handle `InvokeEvent.INVOKE` event in your application https://airsdk.dev/reference/actionscript/3.0/flash/desktop/NativeApplication.html#event:invoke

**Reply by XiaoGen-Taiwan** [reply]
Hello @itlancer , @ajwfrost ,
I'm totally impressed ... thank you soooo much for your help.

I just tried it with macOS and Win11... both are working. 
Great .. and thanks again.

My solution for Win11 using InnoSetup: 
Add following lines to the [Registry] section whereby scheme = 'xxx':


```
; Register custom url scheme<br/>
Root: HKA; Subkey: "Software\Classes\xxx"; ValueType: "string"; ValueData: "URL:Custom Protocol"; Flags: uninsdeletekey
Root: HKA; Subkey: "Software\Classes\xxx"; ValueType: "string"; ValueName: "URL Protocol"; ValueData: ""
Root: HKA; Subkey: "Software\Classes\xxx\DefaultIcon"; ValueType: "string"; ValueData: "{app}\{#MyAppExeName},0"
Root: HKA; Subkey: "Software\Classes\xxx\shell\open\command"; ValueType: "string"; ValueData: """{app}\{#MyAppExeName}"" ""%1"""
```

**Comment by 2jfw** [comment]
**EDIT:** Here the content of a `.reg` file which will register a custom URI scheme `semsy://`

```
Windows Registry Editor Version 5.00

[HKEY_CLASSES_ROOT\semsy]
@="\"URL:Semsy Protocol\""
"URL Protocol"=""

[HKEY_CLASSES_ROOT\semsy\DefaultIcon]
@="\"C:\\Program Files\\Dallmeier\\SeMSy5\\SeMSyDesktop\\SeMSyDesktop.exe\",0"

[HKEY_CLASSES_ROOT\semsy\shell]

[HKEY_CLASSES_ROOT\semsy\shell\open]

[HKEY_CLASSES_ROOT\semsy\shell\open\command]
@="\"C:\\Program Files\\Dallmeier\\SeMSy5\\SeMSyLaunch\\SeMSyLaunch.exe\" \"%1\""

```

**Reply by XiaoGen-Taiwan** [reply]
Thanks @2jfw
